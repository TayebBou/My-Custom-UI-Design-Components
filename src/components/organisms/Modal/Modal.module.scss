@import "../../../global_style_settings/colors";

.modal {
  position: fixed;
  z-index: 501;
  margin: 0 auto;
  background-color: white;
  border-radius: 18px;
  box-shadow: 0px 0px 31px -14px;
  animation: appearModal 0.2s linear forwards;
  left: 25%;
  top: 35vh;
  min-height: 30vh;
  width: 50%;

  @media (orientation: portrait) {
    left: 5%;
    width: 90%;
  }
}

.out-modal {
  animation: disappearModal 0.17s linear forwards;
}

.header {
  margin: 0 auto;
  padding: 1.2em 1.2em 0.6em 0.8em;
  display: flex;
  h1 {
    margin: 0.3em 1em;
    font-size: 1.4em;
    font-weight: 400;
  }
}

.close-button {
  background-color: transparent;
  border: 1px solid $primaryColor;
  border-radius: 5px;
  margin-left: auto;
  width: 35px;
  height: 35px;
  display: flex;
}

.close-button:hover {
  transform: scale(1.1);
  transition: transform 0.3s;
}

.close-icon {
  align-self: center;
  margin: auto;
}

.body::before {
  content: "";
  display: block;
  width: 97%;
  margin: auto;
  border-top: 1px solid $primaryColor;
  padding-bottom: 0.8em;
}

.body {
  padding: 0 1.8em 0.8em 1.8em;
}

.black-font {
  z-index: 500;
  position: fixed;
  width: 100%;
  height: 100vh;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.379);
  animation: appearFont 0.2s linear forwards;
}

.out-black {
  animation: disappearFont 0.17s linear forwards;
}

@keyframes appearFont {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@keyframes disappearFont {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}

@keyframes appearModal {
  0% {
    opacity: 0;
    transform: scale(0);
  }
  100% {
    opacity: 1;
    transform: scale(1);
  }
}

@keyframes disappearModal {
  0% {
    opacity: 1;
    transform: scale(1);
  }
  100% {
    opacity: 0;
    transform: scale(0);
  }
}
