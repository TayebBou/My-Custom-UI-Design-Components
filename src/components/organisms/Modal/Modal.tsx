import { FC, useEffect, useRef, useState } from 'react'
import Button from '../../atoms/Button/Button'
import Icon from '../../atoms/Icon/Icon'
import styles from './Modal.module.scss'
import close from '../../../assets/images/cancel.png'

type ModalProps = {
  onDisplay: boolean
  onClose: () => void
  children: React.ReactNode
  className?: string
  title?: string
}

const Modal: FC<ModalProps> = (props) => {
  const { onDisplay, children, onClose, className, title } = props
  const [exiting, setExiting] = useState(false)
  const timeout = useRef<NodeJS.Timeout | null>(null)

  const handleClose = () => {
    setExiting(true)
    timeout.current = setTimeout(() => {
      onClose()
      setExiting(false)
    }, 170)
  }

  useEffect(() => {
    if (!onDisplay) {
      if (timeout.current !== null) {
        clearTimeout(timeout.current)
      }
    }
  }, [onDisplay])

  return (
    <>
      {onDisplay && (
        <>
          <div
            className={
              styles.modal +
              ' ' +
              className +
              ' ' +
              (exiting ? styles['out-modal'] : '')
            }
          >
            <div className={styles.header}>
              <h1>{title}</h1>
              <Button className={styles['close-button']} onClick={handleClose}>
                <Icon
                  className={styles['close-icon']}
                  alt="close icon"
                  src={close}
                  size="17px"
                />
              </Button>
            </div>
            <div className={styles.body}>{children}</div>
          </div>
          <div
            className={
              styles['black-font'] + ' ' + (exiting ? styles['out-black'] : '')
            }
            onClick={handleClose}
          />
        </>
      )}
    </>
  )
}

Modal.defaultProps = {
  className: '',
  title: '',
}

export default Modal
