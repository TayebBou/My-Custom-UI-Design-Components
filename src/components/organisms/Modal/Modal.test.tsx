// @ts-nocheck

import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Modal from "./Modal";

const clearTimeoutMock = jest.spyOn(global, "clearTimeout");
const onCloseMock = jest.fn();

const props = {
  onDisplay: true,
  onClose: onCloseMock,
  className: "Modal__test",
  title: "Test Modal Title",
};

describe("Modal component", () => {
  test("should render correctly with props", () => {
    // Arrange
    const { container } = render(
      <Modal {...props}>
        <p>This is the modal content.</p>
      </Modal>
    );
    // Assert
    expect(container).toMatchSnapshot();
  });
  test("should render correctly without props", () => {
    // Arrange
    const { container } = render(<Modal />);
    // Assert
    expect(container).toMatchSnapshot();
  });
  test("when user click on the black font the onClose method is called after 0.17s (time for animation)", async () => {
    // Arrange
    render(
      <Modal {...props}>
        <p>This is the modal content.</p>
      </Modal>
    );
    const blackFont = screen.getByTestId("black-font");
    // Act
    fireEvent.click(blackFont);
    // Assert
    await waitFor(() => {
      expect(onCloseMock).toHaveBeenCalled();
    });
  });
  test("when user click on the close button the onClose method is called after 0.17s (time for animation)", async () => {
    // when the modal is with onDisplay prop at false there is no timer to clear
    const { rerender } = render(
      <Modal onDisplay={false} onClose={onCloseMock}>
        <p>This is the modal content.</p>
      </Modal>
    );
    expect(clearTimeoutMock).not.toHaveBeenCalled();
    rerender(
      <Modal {...props}>
        <p>This is the modal content.</p>
      </Modal>
    );
    const closeButton = screen.getByAltText("close icon");
    // Act
    fireEvent.click(closeButton);
    expect(clearTimeoutMock).not.toHaveBeenCalled();
    // Assert
    await waitFor(() => {
      expect(onCloseMock).toHaveBeenCalled();
    });
    clearTimeoutMock.mockClear(); // clear the clearTimeoutMock because waitFor is actually invoking a clearTimeout
    // when the modal is with onDisplay prop at false after the closing there is a timer to clear
    rerender(
      <Modal onDisplay={false} onClose={onCloseMock}>
        <p>This is the modal content.</p>
      </Modal>
    );
    expect(clearTimeoutMock).toHaveBeenCalled();
  });
});
