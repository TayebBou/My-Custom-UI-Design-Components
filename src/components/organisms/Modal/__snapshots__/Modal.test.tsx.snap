// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render correctly with props 1`] = `
<div>
  <div
    class="modal Modal__test "
  >
    <div
      class="header"
    >
      <h1>
        Test Modal Title
      </h1>
      <button
        class="button close-button"
        type="button"
      >
        <img
          alt="close icon"
          class="close-icon"
          src="cancel.png"
          style="width: 17px; height: 17px; filter: none;"
        />
      </button>
    </div>
    <div
      class="body"
    >
      <p>
        This is the modal content.
      </p>
    </div>
  </div>
  <div
    class="black-font "
    data-testid="black-font"
  />
</div>
`;

exports[`Modal component should render correctly without props 1`] = `<div />`;
